Problem with conflicting use statements

This is similar to issue #889 and from the same [BigDFT application](http://bigdft.org/Wiki/index.php?title=BigDFT_website). 

The below code causes flang to emit the following message:
> F90-S-0155-gp is use-associated from modules box and module_defs, and cannot be accessed (big_dft2.f90: 36)

I think that the definition in _module_defs_ should be visible because the definition in _box_ is private to the module and the definition from _Poisson_Solver_ is renamed in the _USE_ statement.

```
module f_precisions
  implicit none
  integer, parameter, public :: f_double = selected_real_kind(15, 307)
end module f_precisions

module box
  use f_precisions, gp=>f_double
  implicit none
  private
end module box

module Poisson_Solver
  use f_precisions
  implicit none
  integer, parameter, public :: gp=f_double
end module Poisson_Solver

module module_defs
  use f_precisions
  implicit none  
  integer, parameter, public :: gp=f_double
end module module_defs

module module_base
  use module_defs
  implicit none  
end module module_base

subroutine PSolver(eh)
  use module_base ! This has a public gp
  ! Removing the next line makes gp visible
  use box ! This has a private gp
  ! Removing the next line makes gp visible
  use Poisson_Solver, except_gp => gp ! This has a public gp
  implicit none
  real(gp), intent(out) :: eh ! Should be able to use gp from module_base
END SUBROUTINE PSolver
```
The reproducer is slightly unstable in that removing either of _box_ or _Poisson_Solver_ _USE_ statements causes flang to compile this with no error. Removing the _USE_ of _module_base_ does correctly cause an error because _gp_ is unknown.

Thanks for your comment, @kiranchandramohan . It was very helpful.

I found that between what you described as point 2 (the private flag set to 1 in `sym_in_scope()` called from `getocsym()`) and point 3 (`sym_in_scope()` called from `refsym()`), the [private bit is cleared](https://github.com/flang-compiler/flang/blob/5755e102155f81bbef525390cab84f655da396a7/tools/flang1/flang1exe/module.c#L506) for `SPTR=636` by the `find_def_in_most_recent_scope()` function called by semant1 (L881)->apply_use_stmts->apply_use.

Actually, when I commented out line 506, which sets the private bit to 0, I got the compilation to pass (or rather - I got it to fail at `undefined reference to `MAIN_'`).

So now we have the exact line to blame, but most likely removing it is not the best idea without understanding why it is being called. Perhaps a better fix would be to store the bit's value and set it back once `use` statement processing is done?

--- a/test/f90_correct/inc/use00.mk
+++ b/test/f90_correct/inc/use00.mk
@@ -1,0 +1,20 @@
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+
+########## Make rule for test use00  ########
+
+build:
+	@echo ------------------------------------ building test $(TEST)
+	-$(CC) -c $(CFLAGS) $(SRC)/check.c -o check.$(OBJX)
+	-$(FC) -c $(FFLAGS) $(LDFLAGS) $(SRC)/$(TEST).f90 -o $(TEST).$(OBJX)
+	-$(FC) $(FFLAGS) $(LDFLAGS) $(TEST).$(OBJX) check.$(OBJX) $(LIBS) -o $(TEST).$(EXESUFFIX)
+
+run:
+	@echo ------------------------------------ executing test $(TEST)
+	./$(TEST).$(EXESUFFIX)
+
+verify: ;
+

--- a/test/f90_correct/lit/use00.sh
+++ b/test/f90_correct/lit/use00.sh
@@ -1,0 +1,9 @@
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Shared lit script for each tests. Run bash commands that run tests with make.
+
+# RUN: KEEP_FILES=%keep FLAGS=%flags TEST_SRC=%s MAKE_FILE_DIR=%S/.. bash %S/runmake | tee %t 
+# RUN: cat %t | FileCheck %S/runmake

--- a/test/f90_correct/src/use00.f90
+++ b/test/f90_correct/src/use00.f90
@@ -1,0 +1,39 @@
+module f_precisions
+  integer, parameter, public :: f_double = selected_real_kind(15, 307)
+end module f_precisions
+
+module box
+  use f_precisions, gp=>f_double
+  private
+  public :: get_gp
+contains
+  subroutine get_gp(out_val)
+    integer, intent(out) :: out_val
+    out_val = gp
+  end subroutine get_gp
+end module box
+
+module Poisson_Solver
+  use f_precisions
+  integer, parameter, public :: gp=f_double
+end module Poisson_Solver
+
+module module_defs
+  use f_precisions
+  integer, parameter, public :: gp=f_double
+end module module_defs
+
+subroutine PSolver(eh)
+  use module_defs ! This has a public gp
+  use box ! This has a private gp
+  use Poisson_Solver, except_gp => gp ! This has a public gp
+  real(gp), intent(out) :: eh ! Should be able to use gp from module_base
+end subroutine PSolver
+
+program main
+  use box
+  use Poisson_Solver
+  integer :: box_gp
+  call get_gp(box_gp)
+  call check(box_gp, gp, 1) ! box's gp and Solver's gp alias the same variable
+end program

--- a/tools/flang1/flang1exe/module.c
+++ b/tools/flang1/flang1exe/module.c
@@ -503,7 +503,6 @@ find_def_in_most_recent_scope(int sptr, int save_sem_scope_level)
     if (NMPTRG(sptr1) != NMPTRG(sptr))
       continue;
     if (STYPEG(sptr1) == ST_ALIAS && aliased_sym_visible(sptr1)) {
-      PRIVATEP(sptr1, 0);
       HIDDENP(SYMLKG(sptr1), 0);
     }
     if (STYPEG(sptr1) == ST_ALIAS) {

